From f72322825dc0f1326d3401e120e06ca74aeebfe3 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Wed, 26 Jun 2019 02:26:59 +0800
Subject: [PATCH 1/4] ACPI / x86: add quirk for VAIO VAI0001 device hotplug

The VAIO VAI0001 device (which is the notebook controller) will be not
present status when booting, but get hot plugged when we initialized the
EC.

Add quirk for enabling device hotplug for VAI0001.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/acpi/Makefile   |  1 +
 drivers/acpi/internal.h |  5 +++++
 drivers/acpi/scan.c     |  1 +
 drivers/acpi/x86/vaio.c | 25 +++++++++++++++++++++++++
 4 files changed, 32 insertions(+)
 create mode 100644 drivers/acpi/x86/vaio.c

diff --git a/drivers/acpi/Makefile b/drivers/acpi/Makefile
index 5d361e4e3405..e9fac3e9267c 100644
--- a/drivers/acpi/Makefile
+++ b/drivers/acpi/Makefile
@@ -54,6 +54,7 @@ acpi-y				+= property.o
 acpi-$(CONFIG_X86)		+= acpi_cmos_rtc.o
 acpi-$(CONFIG_X86)		+= x86/apple.o
 acpi-$(CONFIG_X86)		+= x86/utils.o
+acpi-$(CONFIG_X86)		+= x86/vaio.o
 acpi-$(CONFIG_DEBUG_FS)		+= debugfs.o
 acpi-$(CONFIG_ACPI_NUMA)	+= numa.o
 acpi-$(CONFIG_ACPI_PROCFS_POWER) += cm_sbs.o
diff --git a/drivers/acpi/internal.h b/drivers/acpi/internal.h
index f6157d4d637a..8336b62bbae1 100644
--- a/drivers/acpi/internal.h
+++ b/drivers/acpi/internal.h
@@ -58,6 +58,11 @@ void acpi_cmos_rtc_init(void);
 #else
 static inline void acpi_cmos_rtc_init(void) {}
 #endif
+#ifdef CONFIG_X86
+void acpi_vaio_quirk_init(void);
+#else
+static inline void acpi_vaio_quirk_init(void) {}
+#endif
 int acpi_rev_override_setup(char *str);
 
 void acpi_sysfs_add_hotplug_profile(struct acpi_hotplug_profile *hotplug,
diff --git a/drivers/acpi/scan.c b/drivers/acpi/scan.c
index 0e28270b0fd8..afa9d688d113 100644
--- a/drivers/acpi/scan.c
+++ b/drivers/acpi/scan.c
@@ -2184,6 +2184,7 @@ int __init acpi_scan_init(void)
 	acpi_int340x_thermal_init();
 	acpi_amba_init();
 	acpi_init_lpit();
+	acpi_vaio_quirk_init();
 
 	acpi_scan_add_handler(&generic_device_handler);
 
diff --git a/drivers/acpi/x86/vaio.c b/drivers/acpi/x86/vaio.c
new file mode 100644
index 000000000000..629eeb131387
--- /dev/null
+++ b/drivers/acpi/x86/vaio.c
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * vaio.c - VAIO ACPI quirks
+ * Copyright (C) 2019 Icenowy Zheng <icenowy@aosc.io>
+ */
+
+#include <linux/acpi.h>
+#include <linux/module.h>
+
+#include "../internal.h"
+
+/* The VAI0001 device will hot plugin after we probed the EC */
+static const struct acpi_device_id vaio_quirk_device_ids[] = {
+	{"VAI0001"},
+	{""},
+};
+
+static struct acpi_scan_handler vaio_quirk_handler = {
+	.ids = vaio_quirk_device_ids,
+};
+
+void __init acpi_vaio_quirk_init(void)
+{
+	acpi_scan_add_handler(&vaio_quirk_handler);
+}
-- 
2.21.0

